src/Orleans/Messaging/IncomingMessageBuffer.cs

```diff
@@ -128,6 +131,19 @@ public bool TryDecodeMessage(out Message msg)
             headerLength = 0;
             bodyLength = 0;
 
+            return true;
```

Comment:
Oh, now I understand it, it was not really freeing up memory (unless the buffer is bigger than the max sustained size, which should be unlikely), but it was mostly just re-shuffling. So memory leaks wise we are exactly the same as before.
Perf wise I think the optimization you did for the "opportunistic reset" is great! No unnecessary reshuffling of buffers leads to less locks, and less fragmentation of the array segments, so they will mostly be and stay contiguous.